build: update integration/bazel & @angular/bazel schematics to rules nodejs 0.42.2

gregmagolan opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

gregmagolan committed:
build: update integration/bazel to rules_nodejs 0.42.1

gregmagolan committed:
build: update @angular/bazel schematics to rules_nodejs 0.42.1

gregmagolan committed:
build: update @angular/bazel schematics to use html_insert_assets & pkg_web

IgorMinar commented on integration/bazel/src/BUILD.bazel:
```diff
@@ -64,14 +96,15 @@ terser_minified(
 )
 
-web_package(
+pkg_web(
```
`pkg_web`? that seems to be a very unfortunate name especially when compared to `web_package`. Why this odd rename?

gregmagolan replied:
To be consistent with the rest of the Bazel packaging rules: https://docs.bazel.build/versions/master/be/pkg.html

IgorMinar replied:
sgtm

IgorMinar commented on integration/bazel/src/BUILD.bazel:
```diff
@@ -34,15 +35,46 @@ filegroup(
+html_insert_assets(
+    name = "inject_scripts",
+    # We can't output "index.html" since that collides with the input file.
+    # We output "_/index.html" instead ond remap in ts_devserver & pkg_web
```
typo: ond -> and

gregmagolan replied:
👍 thanks

IgorMinar commented on integration/bazel/src/BUILD.bazel:
```diff
@@ -34,15 +35,46 @@ filegroup(
+        "--out",
+        "$@",
+        "--assets",
+        # We load zone.js outside the bundle. That's because it's a "pollyfill"
```
typo: pollyfill -> polyfill

gregmagolan replied:
👍 thanks

IgorMinar commented on integration/bazel/src/BUILD.bazel:
```diff
@@ -34,15 +35,46 @@ filegroup(
+        # which speculates that such features might be available in a browser.
+        # Also it's tricky to configure dead code elimination to understand that
+        # zone.js is used, given that we don't have any import statement that
+        # imports from it.
```
why don't you import as the first import in `main.ts` (or create polyfill.ts) similar to how the CLI does it: https://github.com/angular/angular-cli/blob/a956cea8673ee04c3054742afedd06743170deda/packages/schematics/angular/application/files/src/polyfills.ts.template#L58

it must the the first import though as we don't want to evaluate any other modules before the zone.js is evaluated.

once you do that, all of this html_insert_asset won't be necessary.

gregmagolan replied:
Sure. That would simplify things here. For the purposes of the integration test the zone.js script & bundle script tags could just go into the source `index.html` itself. The purpose of the integration test is is to test @angular/bazel & ng_module & ng_package so there is no need to exercise `html_insert_assets`.

For bazel-schematics we don't have control over the sources or the html file and `html_insert_assets` is meant to replace the inject script functionality `pkg_web` was take out of pkg_web (mix of concerns there as the packaging rule shouldn't also inject scripts into index.html).

In bazel-schematics we also inject `global_styleseet.css`:

```
html_insert_assets(
    name = "inject_scripts",
    outs = ["_/index.html"],
    args = [
        "--html",
        "$(location :index.html)",
        "--out",
        "$@",
        "--roots",
        "$(RULEDIR)",
        "--assets",
        "$(location :global_stylesheet.css)",
        "$(location @npm//:node_modules/zone.js/dist/zone.min.js)",
        "bundle.min.js",
    ],
    data = [
        ":index.html",
        ":global_stylesheet.css",
        "@npm//:node_modules/zone.js/dist/zone.min.js",
    ],
)
```

In the simple case this is all hidden from users unless the build with `--leaveBazelFilesOnDisk`.

gregmagolan replied:
Simplified. No more `html_insert_assets` in integration/bazel :)

gregmagolan committed:
build: remove html_insert_assets complication from integration/bazel

For the purposes of the integration test the zone.js script & bundle script tags could just go into the source index.html itself. The purpose of the integration test is is to test @angular/bazel & ng_module & ng_package so there is no need to exercise html_insert_assets.

gregmagolan committed:
build: update integration/bazel & @angular/bazel schematics to rules_nodejs 0.42.1

This release brings a bug fix that https://github.com/angular/angular/pull/34243 is waiting on in order to remove rules_nodejs patches: fix(builtin): additional_root_paths in pkg_web should also include paths in genfiles and bin dirs (bazelbuild/rules_nodejs#1402)

josephperrott reviewed (approved):
LGTM

josephperrott commented:
Caretaker Note: Global Approval, @IgorMinar's requested changes were addressed.

kara commented:
@gregmagolan This broke CI tests once it was merged into master. See https://circleci.com/workflow-run/9482b603-90a1-474b-bcf3-eaabbda0fbf5

gregmagolan commented:
Thanks for the head up Kara. Will TAL after BazelCon tonight. About to do a lightning talk so good timing :)

IgorMinar reviewed (approved):
thanks for resolving my comments. please figure out what's up with the windows failure. otherwise lgtm

gregmagolan commented:
Will make a new PR to update to rules_nodejs 1.0.0 instead.
